dm/relay/local_reader.go

```diff
@@ -585,19 +588,25 @@ func (r *BinlogReader) parseFile(
 		}
 
 		// align with MySQL
-		// if an event's gtid has been contained by given gset
-		// replace it with HEARTBEAT event
-		// for Mariadb, it will bee replaced with MARIADB_GTID_LIST_EVENT
-		// In DM, we replace both of them with HEARTBEAT event
-		if state.replaceWithHeartbeat {
+		// ref https://github.com/pingcap/tiflow/issues/5063#issuecomment-1082678211
+		// heartbeat period is implemented in LocalStreamer.GetEvent
+		if state.skipGTID {
 			switch e.Event.(type) {
 			// Only replace transaction event
 			// Other events such as FormatDescriptionEvent, RotateEvent, etc. should be the same as before
-			case *replication.RowsEvent, *replication.QueryEvent, *replication.GTIDEvent, *replication.XIDEvent, *replication.TableMapEvent:
+			case *replication.RowsEvent, *replication.QueryEvent, *replication.GTIDEvent,
+				*replication.MariadbGTIDEvent, *replication.XIDEvent, *replication.TableMapEvent:
 				// replace with heartbeat event
-				e = event.GenHeartbeatEvent(e.Header)
+				state.lastSkipGTIDHeader = e.Header
 			default:
 			}
+			return nil
+		} else if lastSkipGTID && state.lastSkipGTIDHeader != nil {
+			// skipGTID is turned off after this event
```

Comment:
```suggestion
			// lastSkipGTID is turned off after this event
```

Reply:
I read it again and it looks like skipGTID will not be reset because r.advanceCurrentGtidSet(gtidStr) will always return true here 😵

Reply:
return false when met un-synced gtids

Reply:
thanks, I figured it out, now only send heartbeat when a new event comining and last event is skipped